core/google/cloud/future/operation.py

```diff
@@ -0,0 +1,250 @@
+                    'Unknown operation error')
+                self.set_exception(exception)
+
+    def _refresh_and_update(self):
+        """Refresh the operation and update the result if needed."""
+        # If the currently cached operation is done, no need to make another
+        # RPC as it will not change once done.
+        if not self._operation.done:
+            self._operation = self._refresh()
+            self._set_result_from_operation()
+
+        return self._operation
+
+    def done(self):
+        """Checks to see if the operation is complete.
+
+        Returns:
+            bool: True if the operation is complete, False otherwise.
+        """
+        operation = self._refresh_and_update()
+        return operation.done
+
+    def cancel(self):
+        """Attempt to cancel the operation.
+
+        Returns:
+            bool: True if the cancel RPC was made, False if the operation is
+                already complete.
+        """
+        if self.done():
+            return False
+
+        self._cancel()
+        return True
+
+    def cancelled(self):
+        """True if the operation was cancelled."""
+        operation = self._refresh_and_update()
+        return (operation.HasField('error') and
+                operation.error.code == code_pb2.CANCELLED)
+
+
+def _refresh_http(api_request, operation_name):
+    """Refresh an operation using a JSON/HTTP client.
+
+    Args:
+        api_request (Callable): A callable used to make an API request. This
+            should generally be an instance of
+            :meth:`google.cloud._http.Connection.api_request`.
+        operation_name (str): The name of the operation.
+
+    Returns:
+        google.longrunning.operations_pb2.Operation: The operation.
+    """
+    path = 'operations/{}'.format(operation_name)
+    api_response = api_request(method='GET', path=path)
+    return json_format.ParseDict(
+        api_response, operations_pb2.Operation())
+
+
+def _cancel_http(api_request, operation_name):
+    """Cancel an operation using a JSON/HTTP client.
+
+    Args:
+        api_request (Callable): A callable used to make an API request. This
+            should generally be an instance of
+            :meth:`google.cloud._http.Connection.api_request`.
+        operation_name (str): The name of the operation.
+    """
+    path = 'operations/{}:cancel'.format(operation_name)
+    api_request(method='POST', path=path)
+
+
+def from_http_json(operation, api_request, result_type, **kwargs):
+    """Create an operation future from using a HTTP/JSON client.
+
+    This interacts with the long-running operations `service`_ (specific
+    to a given API) vis `HTTP/JSON`_.
+
+    .. _HTTP/JSON: https://cloud.google.com/speech/reference/rest/\
+              v1beta1/operations#Operation
+
+    Args:
+        operation (dict): Operation as a dictionary.
+        api_request (Callable): A callable used to make an API request. This
```

Comment:
@dhermes I think this is the right level to pass in here. The existing `operation` expects you to pass in a `Client` and then proceeds to access `Client._connection.api_request` which seemed beyond icky to me. Let me know if you have any problems with this being the request primitive.